### ISSUE TITLE
Privilege check doesn't work

### ISSUE THREAD

**Issue by jmenashe** [da2x/EdgeDeflector#51]
If the user runs the program without any arguments, it performs an elevated privilege check. If the check fails, the program then requests permissions necessary for registry editing. The program then exits without completing the installation. The `else` clause [found here](https://github.com/da2x/EdgeDeflector/blob/master/EdgeDeflector/Program.cs#L195-L202) is the culprit.

More critically, however, these security permissions are not sufficient to complete the installation (in my experience). I was able to get sufficient permissions by adding them to the app manifest, but this approach has the downside of requiring permissions every time the application is run. I have solved this problem by splitting the program into two projects: an installer and the deflector. The installer uses the elevated privileges and the deflector just rewrites the url requests.

You can see my [forked repository](https://github.com/jmenashe/EdgeDeflector) for my complete solution to these issues. Rather than using an msi installer (I didn't see any code for this in your repository), the user just has to run EdgeDeflectorInstaller.exe. This way, all necessary code from installation to runtime are included in the repository. 

I am not submitting a pull request in case you want to go a different direction with your project, but I wanted to provide my take in case it's helpful. If you do want me to do a pull request just let me know. The source you've provided is already a huge help and I would be happy to contribute however I can. Thanks for putting all of this together!

**Comment by da2x** [comment]
OK, I’m a bit confused. What exactly did you do to make the protocol registry fail? I don’t understand how you ran into these problems? Did you build everything yourself? or did you use the provided installer?

> The else clause [found here](https://github.com/da2x/EdgeDeflector/blob/master/EdgeDeflector/Program.cs#L195-L202) is the culprit.

If not elevated then ask for permissions and relaunch with elevated privileges. If elevated permissions then continue installation (not that the else statement is the path taken by the elevated relaunched process — or when launched from the installer.) It’s not super-neat but it gets the job done.

The registry-registering code was included in the main executable because the code is tiny and it makes it somewhat portable. Notably, users must often relaunch that part of the process manually after Windows update barges into their computers and resets Microsoft defaults.

----

Anyhow, the [existing installer](https://github.com/da2x/EdgeDeflector/blob/master/EdgeDeflector/resources/nsis_installer.nsi) already takes care of everything (note that it uses the main executable to register itself, but then it’s launched with elevated permissions already).

I’m not sure why you’d change it from an exe to an winexe either. There is no console window with normal operation? No one has ever reported seeing a console window.

**Comment by jmenashe** [comment]
I built everything myself and tried to install by running the executable. I did not use your installer because it was not included in the build and there was no way for me to verify the code being used.

If I run the executable that your code produces, it asks for elevated privileges and then exits. This happens every time the executable is run and no install ever takes place. Maybe it doesn't work because you have a separate installer and you don't expect the main executable to install the handler?

**Comment by da2x** [comment]
> If I run the executable that your code produces, it asks for elevated privileges and then exits.

Yes. The process quits and starts anew with higher privileges. That is how process privilege escalation works. Did you check the registry to see if the changes actually took place? or how did you test this?

**Comment by jmenashe** [comment]
> The process quits and starts anew with higher privileges. That is how process privilege escalation works. 

This is not how it works on my system; on my installation, programs must request elevated privileges on every run. While I do believe that your settings have allowed you to circumvent UAC's standard restrictions, this seems to be the norm.

> Did you check the registry to see if the changes actually took place? or how did you test this?

First, I "installed" EdgeDeflector by running the executable with no arguments. Next, I noticed that the EdgeDeflector was never being called. This was my first hint that the installation was unsuccessful. I confirmed my suspicions by stepping through the application with the Visual Studio debugger. I found that on every run the application would request elevated permissions and then exit immediately. Moreover, the permissions it requested were not enough to get around UAC (as I noted above). I ultimately had to put the permissions requirements into the app manifest; to avoid requiring manual user input on every link conversion, I split the application into an "installer" (no arguments) and a "deflector" (1 URI argument).
